security/pkg/k8s/chiron/controller.go

```diff
@@ -0,0 +1,591 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package chiron
+
+import (
+	"bytes"
+	"fmt"
+	"path/filepath"
+	"sync"
+	"time"
+
+	"k8s.io/api/admissionregistration/v1beta1"
+	"k8s.io/client-go/kubernetes"
+
+	"github.com/ghodss/yaml"
+	"github.com/howeyc/fsnotify"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+
+	certclient "k8s.io/client-go/kubernetes/typed/certificates/v1beta1"
+	corev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/client-go/tools/cache"
+
+	"istio.io/istio/security/pkg/listwatch"
+	"istio.io/istio/security/pkg/pki/ca"
+	"istio.io/istio/security/pkg/pki/util"
+	"istio.io/pkg/log"
+)
+
+/* #nosec: disable gas linter */
+const (
+	// The Istio webhook secret annotation type
+	IstioSecretType = "istio.io/webhook-key-and-cert"
+
+	// The ID/name for the certificate chain file.
+	CertChainID = "cert-chain.pem"
+	// The ID/name for the private key file.
+	PrivateKeyID = "key.pem"
+	// The ID/name for the CA root certificate file.
+	RootCertID = "root-cert.pem"
+
+	// For debugging, set the resync period to be a shorter period.
+	secretResyncPeriod = 10 * time.Second
+	// secretResyncPeriod = time.Minute
+
+	recommendedMinGracePeriodRatio = 0.2
+	recommendedMaxGracePeriodRatio = 0.8
+
+	// The size of a private key for a leaf certificate.
+	keySize = 2048
+
+	// The number of retries when requesting to create secret.
+	secretCreationRetry = 3
+
+	// The interval for reading a certificate
+	certReadInterval = 500 * time.Millisecond
+	// The number of tries for reading a certificate
+	maxNumCertRead = 20
+
+	// The delay introduced to debounce the CA cert events
+	watchDebounceDelay = 100 * time.Millisecond
+)
+
+var (
+	// TODO (lei-tang): the secret names, service names, ports may be moved to CLI.
+
+	// WebhookServiceNames is service names of the webhooks.
+	WebhookServiceNames = []string{
+		"istio-sidecar-injector",
+		"istio-galley",
+	}
+
+	// WebhookSecretNames is secret names of the webhooks. Each item corresponds to an item
+	//at the same index in WebhookServiceNames.
+	WebhookSecretNames = []string{
+		"istio.webhook.istio-sidecar-injector-service-account",
```

Comment:
There's no relationship with the service accounts.
Let's define the secrets names to be `istio.webhook.<servicename>`, such as:
`istio.webhook.istio-sidecar-injector` and `istio.webhook.istio-galley`.
You can remove this array.

Reply:
Done.